File: SWEA/D4/1224. ［S／W 문제해결 기본］ 6일차 － 계산기3/［S／W 문제해결 기본］ 6일차 － 계산기3.cpp
#include<iostream>
#include<stack>
using namespace std;

int priority(char oper) {
    if (oper == '+' || oper == '-') return 0;
    if (oper == '*' || oper == '/') return 1;
    else return -1;
}

int operating(int a, int b, char oper) {
    switch (oper)
    {
    case '+':
        return a+b;
        break;
    case '-':
        return a-b;
        break;
    case '*':
        return a*b;
        break;
    case '/':
        return a/b;
        break;
    }
    return 0;
}

string postfix(const string &midfix) {
    string s = "";
    stack<char> oper;
    for (char ch : midfix) {
        if (ch >= '0' and ch <= '9') s.push_back(ch);
        else if (ch == '(') oper.push(ch);
        else if (ch == ')') {
            while (!oper.empty() && oper.top() != '(') {
                s.push_back(oper.top());
                oper.pop();
            }
            oper.pop();
        }
        else {
            while (!oper.empty() && priority(oper.top()) >= priority(ch)) {
                s.push_back(oper.top());
                oper.pop();                
            }
            oper.push(ch);
        }
    }
    while (!oper.empty()) {
        s.push_back(oper.top()); oper.pop();
    }
    return s;
}

int calc (const string &pfix) {
    stack<int> num;
    int ans = 0;
    for (char ch : pfix) {
        if (ch >= '0' && ch <='9') num.push(ch-'0');
        else {
            int n2 = num.top();
            num.pop();
            int n1 = num.top();
            num.pop();
            num.push(operating(n1,n2,ch));
        }
    }
    return num.top();
}



int main(int argc, char** argv)
{
	int T = 10;
	// freopen("input.txt", "r", stdin);

	for(int test_case = 1; test_case <= T; ++test_case)
	{
        int n; string p;
        cin >> n; cin >> p;
        cout << '#' << test_case <<' ' << calc(postfix(p)) << '\n';
	}
	return 0;
}

File: SWEA/D4/1224. ［S／W 문제해결 기본］ 6일차 － 계산기3/［S／W 문제해결 기본］ 6일차 － 계산기3.py
# import sys
from collections import deque

# sys.stdin = open('input.txt', 'r')


def operator_priority(op: str) -> int:
    if op == '+' or op == '-':
        return 0
    elif op == '*' or op == '/':
        return 1
    return -1


def operating(a: int, b: int, op: str) -> int:
    if op == '+':
        return a + b
    elif op == '-':
        return a - b
    elif op == '*':
        return a * b
    else:
        return a // b


def postfix(m_fix: str) -> str:
    lst = []
    stack = deque()

    for ch in m_fix:
        if ch.isdecimal():
            lst.append(ch)
        elif ch =='(':
            stack.append(ch)
        elif ch == ')':
            while stack and stack[-1] != '(':
                lst.append(stack.pop())
            stack.pop()
        else:
            while stack and operator_priority(stack[-1]) >= operator_priority(ch):
                lst.append(stack.pop())
            stack.append(ch)
    while stack:
        lst.append(stack.pop())
    return ''.join(lst)


def calc(p_fix: str) -> int:
    res = 0
    stack = deque()
    for ch in p_fix:
        if ch.isdecimal():
            stack.append(ord(ch) - ord('0'))
        else:
            n2 = stack.pop()
            n1 = stack.pop()
            stack.append(operating(n1, n2, ch))
    return stack[-1]


T = 10
for tc in range(1, T + 1):
    num = int(input())
    inp = input()
    print(f'#{tc} {calc(postfix(inp))}')
